### ISSUE TITLE
a question about deduplicate_bismark

### ISSUE THREAD

**Issue by taojinsheng** [FelixKrueger/Bismark#161]
Hi Felix,

In one of issues about bismark bam deduplication,I noted that you write "from those we work out the strand via the read conversion and genome conversion tags (OT and CTOT are +, OB and CTOB are -)".
However,when I checked the source code of deduplicate_bismark,I noted the following lines:
 my $index;
      if ($read_conversion eq 'CT' and $genome_conversion eq 'CT') { ## original top strand
        $index = 0;
        $strand = '+';
      } elsif ($read_conversion eq 'GA' and $genome_conversion eq 'CT') { ## complementary to original top strand
        $index = 1;
        $strand = '-';
      } elsif ($read_conversion eq 'GA' and $genome_conversion eq 'GA') { ## complementary to original bottom strand
        $index = 2;
        $strand = '+';
      } elsif ($read_conversion eq 'CT' and $genome_conversion eq 'GA') { ## original bottom strand
        $index = 3;
        $strand = '-';
      } 
According to the above codes,OT and CTOB are + and CTOT and OB are -.Is my understanding wrong?I think,even two fragments share the same chromosome,start and end,but if one of them came from Waston genomic strand and the other came from Crick genomic strand,they should not be treated as PCR duplicates.Is my understanding right?I would be greatly appreciated if you could clarify it for me.Many thanks!

Best regards,

Jinsheng

**Comment by FelixKrueger** [comment]
Hi Jinsheng,

Congratulations, you have found the very reason that makes BS-Seq give you a headache! :)

I think the answer to this is indeed two-fold:
1) It is true that OT and CTOT give you methylation information for the + strand (the former top or Watson strand), and OB and CTOB yield methylation information for the - strand (the former bottom or Crick strand).
2) It is however also true that CT and CTOT strands align as forward and reverse reads (or read pairs for paired-end), respectively. This means that - in terms of deduplication - reads that align to the OT and CTOT strands are to be treated separately, even if they hypothetically share the same start coordinates (and end coordinates for paired-end reads). Equally, you could imagine that you also get reads to the exact same position from the CTOB and OB strands, but again they would most likely be all unique reads that have to be treated differently in terms of deduplication.

Luckily most alignments are directional, so you only ever have to deal with OT and OB reads, and the situation is not as complicated is it could be for non-directional data, such as single-cell data.

I hope this helps?

**Comment by taojinsheng** [comment]
Hi Felix,

Thank you for quick response!

I am still a little confused.For directional library preparation,I suppose,if single end,we get reads always from originally bisulfite converted strands,therefore it is OT or OB.Similarly,if paired end,since read 1 always come from the originally bisulfite converted genomic strands,therefore read2 always come from its reverse complement,i.e.,CTOT or CTOB. So,OT/CTOT or OB/CTOB give us the methylation from the originally two bisulfite converted strands(Waston and Crick strand) respectively,I don't know why reads align to the OT and CTOT (or similarly OB and CTOB) strands should be treated seprately.

Yes,I agree that for directional data,the first appearing read in a read pair in bismark bam file (i.e. the read1) will always be OT or OB(its flag is 99 or 83),so it make no difference.

Best,

Jinsheng

**Comment by FelixKrueger** [comment]
SE alignments are indeed either OT or OB only so there is no real ambiguity there.

I believe the confusion comes from the paired-end nature of reads. As you pointed out, for directional paired-end reads, Read 1 always aligns to OT or OB, exactly like for single-end reads. Now while technically read 2 has to be the reverse complement of the the first read and would align in the opposing direction on the complementary strand (if you were to align read 2 in single-end mode), this is already factored in the paired-end alignment.
In other words: directional paired-end reads also only produce OT and OB alignments, but never CTOT or CTOB ones.

In terms of deduplication this means that both reads are taken into account for each alignment, but since the (read 1) alignments are either OT or OB, you will never have to deal with opposing reads at all. 

I hope this makes sense?

**Comment by taojinsheng** [comment]
Hi Felix,

Thank you for your clarification.My understanding is that,for directional paired-end sequencing dataRead 1 produce OT and OB alignments,Read 2 produce CTOT and CTOB alignments.But you said,"directional paired-end reads also only produce OT and OB alignments, but never CTOT or CTOB ones‘’.I wonder how CTOT and CTOB are defined,since in our directional  paired-end bam file produced by bismark,I saw all Read 2 were marked  with "XR:Z:GA" and "XG:Z:CT" (I think they are CTOT alignments) or "XR:Z:GA" and "XG:Z:GA" (I think they are CTOB alignments) tag combination.Sorry to bother you with so many questions, but thank you for your patience.

Best,

Jinsheng

**Comment by FelixKrueger** [comment]
Hi Jinsheng,

`for directional paired-end sequencing data Read 1 produce OT and OB alignments,Read 2 produce CTOT and CTOB alignments.`

Technically, it is correct that Read 1 and Read 2 align in opposing directions on the other strand (Read 2 is by definition (and PCR) the reverse complemented strand of Read 1). This is however not paired-end alignments are scored. For the strand assignment of a paired-end alignment it only matters to which strand Read 1 aligns to. While Read 1 determines where an aligned paired-end fragment starts, Read 2 is used to determine the end of the fragment. Granted, the methylation calling for Read 2 does not do a C/T lookup but a G/A lookup instead, but this is all handled within Bismark itself and you don't have to worry about the specifics.

If you would only look at single-end alignments your assessment of mapping strands would be correct, but for paired-end alignments you need to look at the combination of all three conversions, namely Read1/Read2/Genome, together to determine the alignment strand. Here is a real life example:

```
Number of sequence pairs with unique best (first) alignment came from the Bowtie2 output:
CT/GA/CT:	87265548	((converted) top strand, OT)
GA/CT/CT:	0	(complementary to (converted) top strand, CTOT)
GA/CT/GA:	0	(complementary to (converted) bottom strand, CTOB)
CT/GA/GA:	86210398	((converted) bottom strand, OB)
```

Does this explain it?

**Comment by taojinsheng** [comment]
Hi Felix,

Thank you for clarification.Yes,I now know the OT/CTOT/OB/CTOB definitions.They are determined by the read1/read2/reference conversion combinations.So,for directional paired end sequencing data,CTOT and CTOB strands will never have any paired end reads being assigned to them.For now,I have no questions about directional data.But for nondirectional data,I may still have some confusion.

Imagine now we have a short double strand DNA fragment,it is prepared with nondirectional protocol and then sequenced.since OT,CTOT,OB,CTOB four strands hypothetically have the same start and same end,so in principle deduplicate_bismark will treat OT and CTOT as unique read pairs but will treat OT and CTOB as duplicated read pairs (since their strand:chr:start:end are the same) or similarly will treat OB and CTOT as duplicated read pairs (since their strand:chr:start:end are the same).Finally,the result is that OT and CTOT will be treated as unique read pairs (if they appear in the bam first) and the remaining two (i.e. OB and CTOB) will be treated as duplicates.Is my understanding right?

Best,
Jinsheng

**Comment by FelixKrueger** [comment]
Right, now I get your point! And I think you are right. For directional (or PBAT) libraries the current way of deduplicating paired-end alignments (`strand`:`chr`:`start`:`end`) should be fine.

For non-directional data we should consider adding another parameter, e.g. `strand identity`, for the deduplication to then work with a combination of: `strand`:`chr`:`start`:`end`:`strand identity`? Or even better still, simply replace the `strand` (the `+` or `-`) with `strand identity` altogether so that we don't have to increase the memory footprint even further. Would you agree?

I suppose under normal conditions the case described here would be a negligibly small corner case for non-directional data (of which there isn't that much around), whenever you sequence e.g. both an OT read and a CTOT read that both start and end at the exact same genomic position. But even if this should never really affect anyone, at some point I am sure someone will devise a method that generates exclusively that kind of data (this is usually what happens). If you don't see any obvious flaw in my thinking I would update this next week.

Thanks again for being so persistent with your questions, I admit it took me a while to get the idea...

**Comment by taojinsheng** [comment]
I am very glad that you finally understood my question!It also took me a while to explain my confusion.Yes,I agree to your suggestion.Thank you so much for your patience!

Best,
Jinsheng

**Comment by FelixKrueger** [comment]
I have now tried to add the strand identity for both default mode and UMI mode in this commit: f0dc4a56849165a78c39f94d647899d5cb4caa5c. Works in both single and paired-end mode. You can try it out by cloning the latest development version.

**Comment by taojinsheng** [comment]
Great.Thank you very much for all you help.

Best,
Jinsheng
